internal/webhook/management_webhook.go

```diff
@@ -121,99 +126,71 @@ func checkComponentsRemoval(ctx context.Context, cl client.Client, oldMgmt, newM
 		return nil
 	}
 
-	release := new(kcmv1.Release)
-	if err := cl.Get(ctx, client.ObjectKey{Name: newMgmt.Spec.Release}, release); err != nil {
-		return fmt.Errorf("failed to get Release %s: %w", newMgmt.Spec.Release, err)
-	}
-
-	removedProvidersSet := make(map[string]struct{})
+	var errs error
 	for _, m := range removedComponents {
 		tplRef := m.Template
-		if tplRef == "" {
+		if tplRef == "" && release != nil {
 			tplRef = release.ProviderTemplate(m.Name)
 		}
 
-		// it does not matter if component has been successfully installed
+		if tplRef == "" {
+			continue
+		}
+
 		prTpl := new(kcmv1.ProviderTemplate)
 		if err := cl.Get(ctx, client.ObjectKey{Name: tplRef}, prTpl); err != nil {
 			return fmt.Errorf("failed to get ProviderTemplate %s: %w", tplRef, err)
 		}
 
-		for _, pn := range prTpl.Status.Providers {
-			removedProvidersSet[pn] = struct{}{}
+		inUsedProviders, err := getInUsedProviderWithContracts(ctx, cl, prTpl)
+		if err != nil {
+			return err
```

Comment:
please wrap the error, otherwise it might be hard to understand what has gone south